packages/aws-cdk/lib/cli/activity-printer/base.ts

```diff
@@ -0,0 +1,194 @@
+        break;
+    }
+  }
+
+  public addActivity(activity: StackActivity) {
+    const status = activity.event.ResourceStatus;
+    const hookStatus = activity.event.HookStatus;
+    const hookType = activity.event.HookType;
+    if (!status || !activity.event.LogicalResourceId) {
+      return;
+    }
+
+    this.stackProgress = activity.progress;
+
+    if (status === 'ROLLBACK_IN_PROGRESS' || status === 'UPDATE_ROLLBACK_IN_PROGRESS') {
+      // Only triggered on the stack once we've started doing a rollback
+      this.rollingBack = true;
+    }
+
+    if (status.endsWith('_IN_PROGRESS')) {
+      this.resourcesInProgress[activity.event.LogicalResourceId] = activity;
+    }
+
+    if (stackEventHasErrorMessage(status)) {
+      const isCancelled = (activity.event.ResourceStatusReason ?? '').indexOf('cancelled') > -1;
+
+      // Cancelled is not an interesting failure reason
+      if (!isCancelled) {
+        this.failures.push(activity);
+      }
+    }
+
+    if (status.endsWith('_COMPLETE') || status.endsWith('_FAILED')) {
+      delete this.resourcesInProgress[activity.event.LogicalResourceId];
+    }
+
+    if (
+      hookStatus !== undefined &&
+        hookStatus.endsWith('_COMPLETE_FAILED') &&
+        activity.event.LogicalResourceId !== undefined &&
+        hookType !== undefined
+    ) {
+      if (this.hookFailureMap.has(activity.event.LogicalResourceId)) {
+        this.hookFailureMap.get(activity.event.LogicalResourceId)?.set(hookType, activity.event.HookStatusReason ?? '');
+      } else {
+        this.hookFailureMap.set(activity.event.LogicalResourceId, new Map<string, string>());
+        this.hookFailureMap.get(activity.event.LogicalResourceId)?.set(hookType, activity.event.HookStatusReason ?? '');
+      }
+    }
+  }
+
+  protected failureReason(activity: StackActivity) {
+    const resourceStatusReason = activity.event.ResourceStatusReason ?? '';
+    const logicalResourceId = activity.event.LogicalResourceId ?? '';
+    const hookFailureReasonMap = this.hookFailureMap.get(logicalResourceId);
+
+    if (hookFailureReasonMap !== undefined) {
+      for (const hookType of hookFailureReasonMap.keys()) {
+        if (resourceStatusReason.includes(hookType)) {
+          return resourceStatusReason + ' : ' + hookFailureReasonMap.get(hookType);
+        }
+      }
+    }
+    return resourceStatusReason;
+  }
+
+  public start({ stack }: { stack: CloudFormationStackArtifact}) {
+    this.active = true;
+    this.resourceTypeColumnWidth = maxResourceTypeLength(stack.template),
+    this.scheduleNextTick();
+  }
+
+  public async stop() {
+    // Do a final pass of all received messages
+    await this.tick();
+
+    // disable ticking and clear timeout
+    this.active = false;
+    if (this.tickTimer) {
+      clearTimeout(this.tickTimer);
+    }
+  }
+
+  private scheduleNextTick() {
+    if (!this.active) {
+      return;
+    }
+
+    this.tickTimer = setTimeout(() => void this.tick(), this.updateSleep);
```

Comment:
It seems like we can respond to events immediately as they come in, in `addActivity`. I don't see why we need another timer here, and it leads to compounding latency of 2 polling timers.

Reply:
Done!